SU2_CFD/src/numerics_direct_mean.cpp

```diff
@@ -4716,6 +4728,223 @@ void CAvgGradCorrected_Flow::ComputeResidual(su2double *val_residual, su2double
   
 }
 
+void CAvgGradCorrected_Flow::GetMeanRateOfStrainMatrix(su2double **S_ij)
+{
+  /* --- Calculate the rate of strain tensor, using mean velocity gradients --- */
+
+  if (nDim == 3){
+    S_ij[0][0] = Mean_GradPrimVar[1][0];
+    S_ij[1][1] = Mean_GradPrimVar[2][1];
+    S_ij[2][2] = Mean_GradPrimVar[3][2];
+    S_ij[0][1] = 0.5 * (Mean_GradPrimVar[1][1] + Mean_GradPrimVar[2][0]);
+    S_ij[0][2] = 0.5 * (Mean_GradPrimVar[1][2] + Mean_GradPrimVar[3][0]);
+    S_ij[1][2] = 0.5 * (Mean_GradPrimVar[2][2] + Mean_GradPrimVar[3][1]);
+    S_ij[1][0] = S_ij[0][1];
+    S_ij[2][1] = S_ij[1][2];
+    S_ij[2][0] = S_ij[0][2];
+  }
+  else {
+    S_ij[0][0] = Mean_GradPrimVar[1][0];
+    S_ij[1][1] = Mean_GradPrimVar[2][1];
+    S_ij[2][2] = 0.0;
+    S_ij[0][1] = 0.5 * (Mean_GradPrimVar[1][1] + Mean_GradPrimVar[2][0]);
+    S_ij[0][2] = 0.0;
+    S_ij[1][2] = 0.0;
+    S_ij[1][0] = S_ij[0][1];
+    S_ij[2][1] = S_ij[1][2];
+    S_ij[2][0] = S_ij[0][2];
+
+  }
+}
+
+void CAvgGradCorrected_Flow::SetReynoldsStressMatrix(su2double turb_ke){
+  unsigned short jDim;
+  su2double **S_ij = new su2double* [3];
+  su2double muT = Mean_Eddy_Viscosity;
+  su2double divVel = 0;
+  su2double density;
+  su2double TWO3 = 2.0/3.0;
+  density = Mean_PrimVar[nDim+2];
+
+  for (iDim = 0; iDim < 3; iDim++){
+    S_ij[iDim] = new su2double [3];
+  }
+
+
+  GetMeanRateOfStrainMatrix(S_ij);
+
+  /* --- Using rate of strain matrix, calculate Reynolds stress tensor --- */
+
+  for (iDim = 0; iDim < 3; iDim++){
+    divVel += S_ij[iDim][iDim];
+  }
+
+  for (iDim = 0; iDim < 3; iDim++){
```

Comment:
A number of loops and mem allocations are running up to '3' in the dimension. Is this safe even for 2D cases? We typically run the loops up to nDim so that it is not hard-coded. Please double-check these throughout the PR, just in case.

Reply:
So the reason we need 3 dimensions, regardless of the dimension of the problem, is that we need 3 ordered eigenvalues and eigenvectors (even if one of them is 0). It makes the rest of the code a lot cleaner to just have everything be of dimension 3. 

I checked through the PR and it is consistent throughout, so it shouldn't cause an issue. That being said, if you want these built to nDim, that can be done. It would require more if statements in the code and I don't know which would be more efficient. 

Currently the eigen-decomposition functions are hard coded to work with symmetric tensors of order 3. But I am working on making that general to n order. But I think I would prefer having the tensors be of order 3 regardless.